pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,353 @@
+			"failed to provision namespace '%s'", nsName)
+	}
+
+	namespace := &corev1.Namespace{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: nsName}, namespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to get namespace '%s'", nsName)
+	}
+	if namespace.Labels == nil {
+		namespace.Labels = make(map[string]string)
+	}
+	namespace.Labels["revision"] = tcNamespace.Revision
+	if err := r.client.Update(context.TODO(), namespace); err != nil {
+		return r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to update namespace '%s'", nsName)
+	}
+
+	log.Info("namespace provisioned", "namespace", tcNamespace)
+	if err := r.setStatusProvisioning(nsTmplSet); err != nil {
+		return err
+	}
+
+	// TODO add validation for other objects
+	return nil
+}
+
+func nextMissingNamespace(tcNamespaces []toolchainv1alpha1.Namespace, namespaces []corev1.Namespace, username string) (*toolchainv1alpha1.Namespace, *corev1.Namespace, bool) {
+	for _, tcNamespace := range tcNamespaces {
+		nsName := ToNamespaceName(username, tcNamespace.Type)
+		namespace, found := findNamespace(namespaces, nsName)
+		if found {
+			if namespace.Status.Phase == corev1.NamespaceActive && namespace.GetLabels()["revision"] != tcNamespace.Revision {
+				return &tcNamespace, &namespace, true
+			}
+		} else {
+			return &tcNamespace, nil, true
+		}
+	}
+	return nil, nil, false
+}
+
+func findNamespace(namespaces []corev1.Namespace, namespaceName string) (corev1.Namespace, bool) {
+	for _, ns := range namespaces {
+		if ns.GetName() == namespaceName {
+			return ns, true
+		}
+	}
+	return corev1.Namespace{}, false
+}
+
+// ToNamespaceName returns NamespaceName formed using given userName and nsType
+func ToNamespaceName(userName, nsType string) string {
+	return fmt.Sprintf("%s-%s", userName, nsType)
+}
+
+// error handling methods
+
+func (r *ReconcileNSTemplateSet) wrapErrorWithStatusUpdate(logger logr.Logger, nsTmplSet *toolchainv1alpha1.NSTemplateSet, statusUpdater func(nsTmplSet *toolchainv1alpha1.NSTemplateSet, message string) error, err error, format string, args ...interface{}) error {
+	if err == nil {
+		return nil
+	}
+	if err := statusUpdater(nsTmplSet, err.Error()); err != nil {
+		logger.Error(err, "status update failed")
+	}
+	return errs.Wrapf(err, format, args...)
+}
+
+func (r *ReconcileNSTemplateSet) updateStatusConditions(nsTmplSet *toolchainv1alpha1.NSTemplateSet, newConditions ...toolchainv1alpha1.Condition) error {
+	var updated bool
+	nsTmplSet.Status.Conditions, updated = condition.AddOrUpdateStatusConditions(nsTmplSet.Status.Conditions, newConditions...)
+	if !updated {
+		// Nothing changed
+		return nil
+	}
+	return r.client.Status().Update(context.TODO(), nsTmplSet)
+}
+
+func (r *ReconcileNSTemplateSet) setStatusProvisionFailed(nsTmplSet *toolchainv1alpha1.NSTemplateSet, message string) error {
```

Comment:
those 5 funcs below look very similar. Can't we factorize the code?

Reply:
same as UserAccCtrl https://github.com/codeready-toolchain/member-operator/blob/master/pkg/controller/useraccount/useraccount_controller.go#L328 but we can revisit though.

Reply:
Yes, it's not easy to make it generic. I did something in MUR controller eg here:
https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/masteruserrecord/masteruserrecord_controller.go#L294
or here
https://github.com/codeready-toolchain/host-operator/blob/master/pkg/controller/masteruserrecord/masteruserrecord_controller.go#L225
but I wanted to make even more generic solution that could be used in all controllers, however, I haven't had time to write it yet.

Reply:
Code re-usage is good but it's not always the best thing to do. We should also care about code readability, how easy it is to use and maintain, etc. We should try to keep a good balance.
And this is the case when I think a little copy-paste is better than a more generic approach which would make code harder to understand (==harder to maintain, == harder to use).